qt: Disable requests context menu actions when appropriate

jarolrod opened the pull request:
The recent requests table will allow you to copy data points even if they do not exist. This PR implements checks to disable the `copy label`, `copy message`, and `copy amount` context menu actions if the respective fields are empty. This brings the recent requests table context menu behavior closer to the behavior seen in the transaction view.

On a payment request entry which does not have a value for label, message, or amount:
| Master        | PR               |
| ----------- | ----------- |
|<img width="169" alt="Screen Shot 2021-02-19 at 1 22 28 AM" src="https://user-images.githubusercontent.com/23396902/108466086-167adc00-7251-11eb-8bd6-13984042bdb3.png">| <img width="169" alt="Screen Shot 2021-02-19 at 1 21 49 AM" src="https://user-images.githubusercontent.com/23396902/108466185-3e6a3f80-7251-11eb-9dd8-492ed07fd638.png">|

`copy URI` never needs to be disabled as an entry in the recent requests table must have a URI even if it doesn't have a label, message, or amount. #213 will add a `copy address` context menu action. This also does not need a check as an entry must be associated with an address.

Below are some more examples of how this PR will behave:
**Has Label, Message, and Amount**
<img width="780" alt="Screen Shot 2021-02-19 at 12 05 38 AM" src="https://user-images.githubusercontent.com/23396902/108466507-c18b9580-7251-11eb-8875-f3aeb9c4c8e9.png">

**Has Label and Amount, but no Message**
<img width="780" alt="Screen Shot 2021-02-19 at 12 05 58 AM" src="https://user-images.githubusercontent.com/23396902/108466421-9b65f580-7251-11eb-97eb-a3bfaa21fa7d.png">

jonasschnelli commented:
utACK 69915b80ccf853b66345b3d612d24eddd3411a2a - I guess comparing against zero for an "empty" amount is acceptable here.

jarolrod commented:
> utACK [69915b8](https://github.com/bitcoin-core/gui/commit/69915b80ccf853b66345b3d612d24eddd3411a2a) - I guess comparing against zero for an "empty" amount is acceptable here.

Comparing against zero is what's used to determine `(no amount requested)` in [recentrequeststablemodel.cpp](https://github.com/bitcoin-core/gui/blob/master/src/qt/recentrequeststablemodel.cpp) Line 84

Talkless commented:
Concept ACK.

Talkless commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -267,9 +267,18 @@ void ReceiveCoinsDialog::copyColumnToClipboard(int column)
+    // disable context menu actions when appropriate
+    const RecentRequestsTableModel * const submodel = model->getRecentRequestsTableModel();
+    const RecentRequestEntry *req = &submodel->entry(sel.row());
+    copyLabelAction->setEnabled(!req->recipient.label.isEmpty());
```
nit: negations could be avoided (simplified) by using `setDisabled` instead (same for all three similar lines).

jarolrod replied:
addressed in 336a643

Talkless commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -267,9 +267,18 @@ void ReceiveCoinsDialog::copyColumnToClipboard(int column)
 void ReceiveCoinsDialog::showMenu(const QPoint &point)
 {
-    if (!selectedRow().isValid()) {
+    QModelIndex sel = selectedRow();
```
nit: `sel` Could be const.

jarolrod replied:
addressed in 336a643

Talkless commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -267,9 +267,18 @@ void ReceiveCoinsDialog::copyColumnToClipboard(int column)
+
+    // disable context menu actions when appropriate
+    const RecentRequestsTableModel * const submodel = model->getRecentRequestsTableModel();
+    const RecentRequestEntry *req = &submodel->entry(sel.row());
```
nit: Pointer `req` is not change within function, could be  `const RecentRequestEntry * const req` (same as `submodel` above).

jarolrod replied:
addressed in 336a643

jarolrod commented:
updated 69915b8 -> 336a643

Addressed @Talkless suggestions, thanks for the review!

Changes:
- variables that can be `const` are made `const`
- refactor context menu action disable logic from using `setEnabled` with negations to using `setDisabled` without negations

MarcoFalke commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -267,9 +267,18 @@ void ReceiveCoinsDialog::copyColumnToClipboard(int column)
+
+    // disable context menu actions when appropriate
+    const RecentRequestsTableModel * const submodel = model->getRecentRequestsTableModel();
+    const RecentRequestEntry * const req = &submodel->entry(sel.row());
```
wouldn't a const reference make more sense?
```suggestion
    const RecentRequestEntry& req = submodel->entry(sel.row());
```

jarolrod replied:
addressed in 3f83f05f52be8f3bac2267d447b5037409cde17d

MarcoFalke reviewed (commented):
tested ACK 336a6435965dc813bdc2bf11398bce564e52c984

didn't review code

hebasto commented:
Concept ACK.

hebasto commented on src/qt/receivecoinsdialog.h:
```diff
@@ -54,6 +54,9 @@ public Q_SLOTS:
     GUIUtil::TableViewLastColumnResizingFixer *columnResizingFixer;
     WalletModel *model;
     QMenu *contextMenu;
+    QAction *copyLabelAction;
+    QAction *copyMessageAction;
+    QAction *copyAmountAction;
```
style nit (see [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy)):
```suggestion
    QAction* copyLabelAction;
    QAction* copyMessageAction;
    QAction* copyAmountAction;
```

jarolrod replied:
Addressed in 3f83f05f52be8f3bac2267d447b5037409cde17d

hebasto commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -267,9 +267,18 @@ void ReceiveCoinsDialog::copyColumnToClipboard(int column)
     }
+
+    // disable context menu actions when appropriate
+    const RecentRequestsTableModel * const submodel = model->getRecentRequestsTableModel();
```
style nit (see [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy)):
```suggestion
    const RecentRequestsTableModel* const submodel = model->getRecentRequestsTableModel();
```

hebasto reviewed (approved):
ACK 336a6435965dc813bdc2bf11398bce564e52c984, tested on Linux Mint 20.1 (Qt 5.12.8).

jarolrod commented:
updated from 336a643  -> 3f83f05f52be8f3bac2267d447b5037409cde17d, thanks @hebasto @MarcoFalke for the review

Changes:
- Adhere to [clang-format](https://github.com/bitcoin/bitcoin/blob/master/src/.clang-format)
- `req` is made to be a const reference as that makes more sense

hebasto commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -266,9 +266,18 @@ void ReceiveCoinsDialog::copyColumnToClipboard(int column)
     }
+
+    // disable context menu actions when appropriate
+    const RecentRequestsTableModel* const submodel = model->getRecentRequestsTableModel();
+    const RecentRequestEntry& req = submodel->entry(sel.row());
+    copyLabelAction->setDisabled(req.recipient.label.isEmpty());
+    copyMessageAction->setDisabled(req.recipient.message.isEmpty());
+    copyAmountAction->setDisabled(req.recipient.amount == 0);
```
```suggestion
    const auto& r = model->getRecentRequestsTableModel()->entry(sel.row()).recipient;
    copyLabelAction->setDisabled(r.label.isEmpty());
    copyMessageAction->setDisabled(r.message.isEmpty());
    copyAmountAction->setDisabled(r.amount == 0);
```

jarolrod replied:
Since 213 was merged with the style used in this PR, I'll close this for consistency

hebasto reviewed (approved):
re-ACK 3f83f05f52be8f3bac2267d447b5037409cde17d, only suggested changes and rebased since my [previous](https://github.com/bitcoin-core/gui/pull/214#pullrequestreview-595579186) review, verified with
```
$ git range-diff master 336a6435965dc813bdc2bf11398bce564e52c984 3f83f05f52be8f3bac2267d447b5037409cde17d
```

jarolrod commented:
Updated from 3f83f05 -> bb3da8f

Changes: Resolve conflict with #213

hebasto reviewed (approved):
re-ACK bb3da8fe410285887f22679c6f08a1f40bcfac8f

luke-jr reviewed (approved):
utACK
